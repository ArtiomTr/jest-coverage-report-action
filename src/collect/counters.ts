import { FileCoverage } from '../typings/JsonReport';

export const standardTotalCounter = (key: keyof FileCoverage) => (
    value: FileCoverage
) => Object.values(value[key]).length;

export const standardCoveredCounter = (key: keyof FileCoverage) => (
    value: FileCoverage
) => Object.values(value[key]).filter((hits) => hits > 0).length;

export const totalBranchesCounter = (value: FileCoverage) =>
    Object.values(value.b).reduce((acc, branch) => acc + branch.length, 0);

export const coveredBranchesCounter = (value: FileCoverage) =>
    Object.values(value.b).reduce(
        (acc, branch) => acc + branch.filter((hits) => hits > 0).length,
        0
    );

const lineDiff = (startLine: number, endLine: number, lastEndLine: number) =>
    Math.max(endLine - Math.max(startLine, lastEndLine), 0);

export const totalLinesCounter = (value: FileCoverage) => {
    let lastEndLine = 0;

    return Object.values(value.statementMap).reduce((acc, statement) => {
        const newLines = lineDiff(
            statement.start.line,
            statement.end.line,
            lastEndLine
        );

        lastEndLine = Math.max(statement.end.line, lastEndLine);

        return acc + newLines;
    }, 0);
};

export const coveredLinesCounter = (value: FileCoverage) => {
    let lastEndLine = 0;

    const totalLines = totalLinesCounter(value);

    const notCoveredLines = Object.entries(value.statementMap).reduce(
        (acc, [key, statement]) => {
            if (value.s[+key] <= 0) {
                const newLines = lineDiff(
                    statement.start.line,
                    statement.end.line,
                    lastEndLine
                );
                lastEndLine = Math.max(statement.end.line, lastEndLine);

                return acc + newLines;
            }

            return acc;
        },
        0
    );

    return Math.max(0, totalLines - notCoveredLines);
};
