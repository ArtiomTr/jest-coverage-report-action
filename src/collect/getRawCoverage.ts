import { join } from 'path';

import { exec } from '@actions/exec';
import { readFile, rm } from 'fs-extra';

import { REPORT_PATH } from '../constants/REPORT_PATH';
import { FailReason } from '../typings/Report';

const joinPaths = (...segments: Array<string | undefined>) =>
    join(...(segments as string[]).filter((segment) => segment !== undefined));

export const getRawCoverage = async (
    testCommand: string,
    branch?: string,
    workingDirectory?: string
): Promise<
    | string
    | { success: false; failReason: FailReason.TESTS_FAILED; error?: Error }
> => {
    if (branch) {
        // NOTE: It is possible that the 'git fetch -all' command will fail due to different file permissions, so allow that to fail gracefully
        try {
            await exec(`git fetch --all --depth=1`);
        } catch (err) {
            console.warn('Error fetching git repository', err);
        }
        await exec(`git checkout -f ${branch}`);
    }

    // NOTE: The `npm ci` command is not used. Because if your version of npm is old, the generated `package-lock.json` will also be old, and the latest version of `npm ci` will fail.
    await rm(joinPaths(workingDirectory, 'node_modules'), {
        recursive: true,
        force: true,
    });

    await exec('npm i', undefined, {
        cwd: workingDirectory,
    });

    let executionError: Error | undefined = undefined;

    try {
        await exec(testCommand, [], {
            cwd: workingDirectory,
        });
    } catch (error) {
        console.error('Test execution failed with error:', error);
        executionError = error instanceof Error ? error : undefined;
    }

    try {
        const outBuff = await readFile(
            joinPaths(workingDirectory, REPORT_PATH)
        );
        return outBuff.toString();
    } catch (error) {
        console.error(
            'Could not read report file located at',
            joinPaths(workingDirectory, REPORT_PATH),
            error
        );

        return {
            success: false,
            failReason: FailReason.TESTS_FAILED,
            error: executionError,
        };
    }
};
