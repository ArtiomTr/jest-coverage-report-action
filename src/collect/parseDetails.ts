import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from './counters';
import { findCommonPath } from './findCommonPath';
import { getPercents } from './getPercents';
import { CoverageDetailsMap } from '../typings/Coverage';
import { JsonReport } from '../typings/JsonReport';

export const parseDetails = (jsonReport: JsonReport) => {
    // Find common root directory
    let trimPath = 0;
    const filepaths = Object.keys(jsonReport.coverageMap);
    if (filepaths.length) {
        const commonRootPath = findCommonPath(filepaths);
        trimPath = commonRootPath.length;
    }

    return Object.entries(jsonReport.coverageMap).reduce<CoverageDetailsMap>(
        (acc, [filename, fileCoverage]) => {
            const normalizedFilename = filename.substr(trimPath);
            acc[normalizedFilename] = {
                filename: normalizedFilename,
                statements: getPercents(
                    standardCoveredCounter('s')(fileCoverage),
                    standardTotalCounter('s')(fileCoverage)
                ),
                functions: getPercents(
                    standardCoveredCounter('f')(fileCoverage),
                    standardTotalCounter('f')(fileCoverage)
                ),
                branches: getPercents(
                    coveredBranchesCounter(fileCoverage),
                    totalBranchesCounter(fileCoverage)
                ),
                lines: getPercents(
                    coveredLinesCounter(fileCoverage),
                    totalLinesCounter(fileCoverage)
                ),
            };
            return acc;
        },
        {}
    );
};
