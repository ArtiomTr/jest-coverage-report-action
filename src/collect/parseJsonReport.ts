import { JsonReport } from '../typings/JsonReport';
import { FailReason } from '../typings/Report';

export const parseJsonReport = (source: string) => {
    try {
        const jsonReport: JsonReport = JSON.parse(source);

        return jsonReport;
    } catch (err) {
        return {
            success: false,
            error: err as Error,
            failReason: FailReason.INVALID_COVERAGE_FORMAT,
        };
    }
};
