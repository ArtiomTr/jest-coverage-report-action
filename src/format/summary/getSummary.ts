import { CoverageSummary } from '../../typings/Coverage';
import { CoverageMap, FileCoverage } from '../../typings/JsonReport';
import { getPercents } from '../getPercents';

export const getSummary = (
    map: CoverageMap,
    totalCounter: (value: FileCoverage) => number,
    coveredCounter: (value: FileCoverage) => number,
    title: string
): CoverageSummary => {
    const total = Object.values(map).reduce(
        (acc, currValue) => acc + totalCounter(currValue),
        0
    );

    const covered = Object.values(map).reduce(
        (acc, currValue) => acc + coveredCounter(currValue),
        0
    );

    return {
        title,
        total,
        covered,
        percentage: getPercents(covered, total),
    };
};
