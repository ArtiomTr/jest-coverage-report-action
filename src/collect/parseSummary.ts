import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from './counters';
import { getSummary } from './getSummary';
import { JsonReport } from '../typings/JsonReport';

export const parseSummary = (jsonReport: JsonReport) => {
    return [
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('s'),
            standardCoveredCounter('s'),
            'Statements'
        ),
        getSummary(
            jsonReport.coverageMap,
            totalBranchesCounter,
            coveredBranchesCounter,
            'Branches'
        ),
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('f'),
            standardCoveredCounter('f'),
            'Functions'
        ),
        getSummary(
            jsonReport.coverageMap,
            totalLinesCounter,
            coveredLinesCounter,
            'Lines'
        ),
    ];
};
