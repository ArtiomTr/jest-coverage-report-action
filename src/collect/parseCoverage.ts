import { parseDetails } from './parseDetails';
import { parseSummary } from './parseSummary';
import { JsonReport } from '../typings/JsonReport';
import { Report } from '../typings/Report';

export const parseCoverage = (jsonReport: JsonReport): Report => {
    return {
        success: true,
        summary: parseSummary(jsonReport),
        details: parseDetails(jsonReport),
    };
};
