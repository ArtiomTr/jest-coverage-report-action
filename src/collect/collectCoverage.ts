import { getRawCoverage } from './getRawCoverage';
import { parseCoverage } from './parseCoverage';
import { parseJsonReport } from './parseJsonReport';
import { JsonReport } from '../typings/JsonReport';
import { Report } from '../typings/Report';

export const collectCoverage = async (
    testCommand: string,
    branch?: string,
    workingDirectory?: string
): Promise<[Report, JsonReport | undefined]> => {
    const source = await getRawCoverage(testCommand, branch, workingDirectory);

    if (typeof source !== 'string') {
        return [source, undefined];
    }

    const jsonReport = parseJsonReport(source);

    if (
        jsonReport.success === false &&
        (jsonReport as { failReason: unknown }).failReason !== undefined
    ) {
        return [jsonReport, undefined];
    }

    return [parseCoverage(jsonReport as JsonReport), jsonReport as JsonReport];
};
