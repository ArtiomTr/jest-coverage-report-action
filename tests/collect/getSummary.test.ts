import jsonReport from './jsonReport.json';
import {
    coveredLinesCounter,
    totalLinesCounter,
} from '../../src/collect/counters';
import { getSummary } from '../../src/collect/getSummary';

describe('getSummary', () => {
    it('Should generate a line number coverage report', () => {
        expect(
            getSummary(
                jsonReport.coverageMap,
                totalLinesCounter,
                coveredLinesCounter,
                'Lines'
            )
        ).toMatchSnapshot();
    });
});
