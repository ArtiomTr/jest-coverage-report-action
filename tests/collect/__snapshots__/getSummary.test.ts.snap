// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSummary Should generate a line number coverage report 1`] = `
Object {
  "covered": 25,
  "percentage": 80.64516129032258,
  "title": "Lines",
  "total": 31,
}
`;
